### ISSUE TITLE
`explore` ideas, bugs, potential changes

### ISSUE THREAD

**Issue by fdncred** [nushell/nushell#7332]
### Related problem

First and foremost, `explore` is awesome!!!! 🥳 🎉 💥.

What follows are ideas collected from Discord. Some are bugs, some are feature request, some are just ideas to consider.

### Describe the solution you'd like

General Changes:

- [ ] `record mode` - We'd like to have a record mode `:record` that activates automatically when the input from nushell is a `record`. The record mode should look similar to the way it does in nushell.
This is an example of how a record looks in nushell
![image](https://user-images.githubusercontent.com/343840/205352806-e1dfce3f-c126-42d1-aa57-1a9c5e35c66a.png)
Can we reproduce that look and feel inside of the tui. It's really similar to the `t` transpose view but we don't want the columns. One way to do this is to hide the columns as if you typed `sys | explore --head false` and automatically transpose the data since we know that it's a record. We'd like to be able to activate this mode from anywhere, if applicable, with `:record`.

- [x] `table mode` - We like this mode but we'd like to make the `i` inspect/cursor functionality automatic some way. We'd like to separate this tui from VIM and having to press `i` to activate some type of additional functionality seems intuitive to VIM users and foreign to other users. We're open to ideas here. It would also be good to activate this mode from anywhere, if applicable, with `:table`.
![image](https://user-images.githubusercontent.com/343840/205354390-4222f70b-0037-48db-aa94-f1ac66e27600.png)

- [x] `expand mode` - Also a functionality in `table` mode, when you execute a command like this `open Cargo.toml | explore`, you can then do `:preview` which gives you the `table -e` look and feel. In this sense, I think it would be better to call this `:expand` or `:expanded`. It would be good to have this functionality available from anywhere, if applicable, with `:expand` or `:e`.
![image](https://user-images.githubusercontent.com/343840/205359776-e019f745-a9d9-4160-bf8b-f6b5697fc61f.png)

- [x] `table mode` - Another example is `open Cargo.toml | expand`, then type `:preview` or newly named `:expand` to go in to that mode. Now, how do I get back to the original tablular view? I'd like to type `:table` to do that. This is an example of being about to hit `:table` from another view and it put you in that mode.

- [ ] `text mode` - This is activated when you do something like `open -r Cargo.toml | explore`. I kind of get a little confused here because I think this is `:preview` mode too (let me know if I'm right or wrong). I think it would be maybe more intuitive to call this `:text`. This is the traditional paging up down left right mode and we love this. It would be good to activate this mode from anywhere with `:text` if possible.
![image](https://user-images.githubusercontent.com/343840/205354641-e97bdef9-f474-4aa9-8b77-0171d2a76765.png)

- [x] `try mode` - This is activated like `explore`, hit enter, then type `:try`. We love this mode ❤️. Although sometimes it's a little tricky to switch between top pane, for input, and bottom pane, for output. I've only been able to do this with up and down arrow keys. I'm wondering if we can also use the `tab` key to switch between them? We'd like try mode to be available from anywhere by typing `:try`, which it may already be.
![image](https://user-images.githubusercontent.com/343840/205356536-2b5d424c-f610-40c7-8e16-3788056ab287.png)

- [x] There's also a bug report in try mode https://github.com/nushell/nushell/issues/7323

- [x] - `help mode` - This is activated with `:help`. It's great but we'd like to make it more discoverable. Could we put something like "For help type :help` in the status bar of any screen?

- [x] - `help mode` - Another thing we'd like to change is how the `:` commands are displayed. It would be nice to just have them all listed instead of having to navigate to another mode and window. Do you think we could do that? Of course, we'd like this screen to be available from anywhere by typeing `:help` or `:h` (i think it may be already)
![image](https://user-images.githubusercontent.com/343840/205356856-655a87b0-d4ef-47aa-9ecb-9e44ff4e5941.png)

- [x] - `nu mode` - Again, yet another mode to love. Activated with `:nu <command` like `:nu ls | sort-by name`. Issue https://github.com/nushell/nushell/issues/7323 occurs here too.

- [x] - `search mode` - Activated with `/` and typing a search string then hitting `enter`. This works according to plan but there's not really a way to get out of it. Example: Can't get this to work right. `open -r cargo.lock` then search for adler like `/adler` and hit `enter`. Now you can go next with `n`. That all works. Now if you want to do something else like try `:preview`, it doesn't work. It's like it's stuck in search mode. Not sure how to turn off search mode - i tried typing `/` and then deleting the search term and the slash and it kind of works, but still seems stuck.

- [x] - general bug report - just logging here too - `explore` in `explore` = bad - maybe should be ignored like `clear` https://github.com/nushell/nushell/issues/7328

- [ ] - documentation - We'd like some documentation on what these configuration settings actually do. Some seem intuitive while others are harder to understand. Maybe we could start with a `# description` at the end of each line that kind of describes what each item is for?
```
    explore: {
        highlight: { bg: 'yellow', fg: 'black' }
        status_bar: { bg: '#C4C9C6', fg: '#1D1F21' }
        command_bar: { fg: '#C4C9C6' }
        split_line: '#404040'
        cursor: true
        # selected_column: 'blue'
        # selected_row: { fg: 'yellow', bg: '#C1C2A3' }
        # selected_cell: { fg: 'white', bg: '#777777' }
        # line_shift: false,
        # line_index: false,
        # line_head_top: false,
        # line_head_bottom: false,
      }
```
I've tried the `line_*` items with true and false and i'm not sure what they do. Maybe they only work in certain modes?

- [x] - `esc` - We'd like to add escape as a way to quit once you get all the way back to the normal view. So, if you're in the first view/root view and you hit the `esc` key, it should quit. We still want to keep `:quit` and `:q` so you can quit from anywhere, but `esc` can maybe do that too at the right time.

One idea that I've had, that can apply to some of the functions above is to use a `helix`/`hx` UX. If you'd played with the `helix` text editor you know that if you hit the `spacebar` it gives you a menu to traverse. I could see this type of thing working in `table` mode and perhaps others or for general navigation somehow. This is what the `helix` editor looks like after hitting `spacebar`.
![image](https://user-images.githubusercontent.com/343840/205354280-b1cad8ec-4fe6-486c-8a0c-e26e11c38877.png)

### Additional context and details

/cc @zhiburt

**Comment by kubouch** [comment]
Also, @zhiburt it would be easier if you could join our Discord. Have you considered it? We love your TUI additions to Nushell and I think the communication would be easier over there rather than ping-ponging with GH issues.

**Comment by webbedspace** [comment]
Seconding the `esc` request.

In addition to `explore` inside `explore`, `exit` inside `explore` should maybe just quit `explore` instead of exiting Nushell.

**Comment by webbedspace** [comment]
Other things:

* Since `try` is now a Nushell command, I think `:try` should possibly be renamed to something else. Maybe `:edit` (since you're essentially "editing" the data)
* `try` mode seems useful enough that I kind of want it to be split off into a different command
* Do those commands *really* need the Vi `:` glyph There's so few of them that it seems a little presumptuous. `n` collides with `nu` and `t` collides with `try`, but those could be worked through. (`n` could maybe be moved to `g` (as in Ctrl-G) or `nu` could be renamed to `do`)

**Comment by zhiburt** [comment]
> * One way to do this is to hide the columns as if you typed `sys | explore --head false` and automatically transpose the data since we know that it's a record.

Made it as you've described.
Though...

> We'd like to be able to activate this mode from anywhere, if applicable, with `:record`.

Aaaaa not sure it these 2 can work together.

Personally I would not do it automatically,
But rather give a use a command to hide column names so it could be done by pressing `<t> + :hide column` something like this.

>  table mode - We like this mode but we'd like to make the i inspect/cursor functionality automatic some way. We'd like to separate this tui from VIM and having to press i to activate some type of additional functionality seems intuitive to VIM users and foreign to other users. We're open to ideas here.

Made it open by hitting <enter> too.

> It would also be good to activate this mode from anywhere, if applicable, with :table.

Made this command available.

> expand mode - Also a functionality in table mode, when you execute a command like this open Cargo.toml | explore, you can then do :preview which gives you the table -e look and feel. In this sense, I think it would be better to call this :expand or :expanded. It would be good to have this functionality available from anywhere, if applicable, with :expand or :e.

Renamed `:preview` to `:expand`.
Added an alias `:e`

>  text mode - This is activated when you do something like open -r Cargo.toml | explore. I kind of get a little confused here.
>  I think it would be maybe more intuitive to call this :text.

That's what I was saying a bunch of times.
I would not do it automatically.

So what I shall do here?

>  because I think this is :preview mode too (let me know if I'm right or wrong).

It is

> I'm wondering if we can also use the tab key to switch between them?

I like `<tab>`. (it was added)

> help mode - This is activated with :help. It's great but we'd like to make it more discoverable. Could we put something like "For help type :help` in the status bar of any screen?

Added the message when first opened.

> help mode - Another thing we'd like to change is how the : commands are displayed. It would be nice to just have them all listed instead of having to navigate to another mode and window. 

We could do something like `help --commands` or whatever like that, to show the commands immediately.

> - search mode - Activated with / and typing a search string then hitting enter. This works according to plan but there's not really a way to get out of it. Example: Can't get this to work right. open -r cargo.lock then search for adler like /adler and hit enter. Now you can go next with n. That all works. Now if you want to do something else like try :preview, it doesn't work. It's like it's stuck in search mode. Not sure how to turn off search mode - i tried typing / and then deleting the search term and the slash and it kind of works, but still seems stuck.

Interesting...
Could you provide a reproducible?
As I believe it works for me.

**Comment by webbedspace** [comment]
>Ahhhh not sure whether it's good?

Think of it like this: vi is an editor. This isn't. Quitting vi unexpectedly is much more impactful than quitting this. Therefore, it makes sense to opt to make quitting easier than vi.

**Comment by fdncred** [comment]
I tried to describe some of these things in #7339 

> Aaaaa not sure it these 2 can work together.

Here's the idea. If you are in table mode and do `:record` it takes column names and makes them keys. If you are in record mode and do `:table` it takes keys and makes columns out of them.

> Personally I would not do it automatically,

I tend to lean this way too but mine is not the only vote.

> But rather give a use a command to hide column names so it could be done by pressing <t> + :hide column something like this.

`:hide #` sounds interesting. we can do that later maybe?

> That's what I was saying a bunch of times.
> I would not do it automatically.
>
> So what I shall do here?

For my example, it may be easy to understand. it's knowing that the extension being opened/displayed is a format nushell supports like .toml, .json, .csv, (any of the `from *` commands, and doing `open` instead of `open -r`.

> We could do something like help --commands or whatever like that, to show the commands immediately.

I'd just rather have your existing table but don't have any of the commands nested in another table. Just have a list of all the keybindings. I looked at your PR and see that you started to change this but I'm not really sure how to make it work where i can see all the keybindings.

> Ahhhh not sure whether it's good?

I'm not sure it's good either. But like I said before, mine is not the only vote. The core-team seems to want to hit esc and quit. IMO, we may end up reversing that decision because I expect people to complain.

> We could do it.
> But maybe it's unnecessary?

I'm not sure either. It's just an idea for us to remember in case we get to a point where we need a different type of navigation. Helix makes good use of these menus in it's editor.

**Comment by amtoine** [comment]
hello there :wave:

first of all...
#### YEAH, this feature is and will be even more awesome when more mature :heart: 

## some thoughts
- first, i wanted to support what has been said by @fdncred :yum: 
- the `explore` command is probably not ready as it is part of the `0.72.1` version and not the `0.72.0` release :thinking: 
for now, i would say i feel quite lost in `explore`... :confused: 
here are some things that i personnally would find super helpful :relieved: 
  - more `vim`-bindings and especially motions
  - the `helix` example is really cool, i love that feature in `helix`, it makes the application easier to work with thanks to the "whichkey" menus and it becomes natural and intuitive to learn the commands and the bindings :grey_exclamation:
some other editor distributions do that as well: `doom emacs`, `spacevim` and `lunarvim` i think
  - and finally a page about the `explore` tool would be great, for instance in the book :+1: 
  
this is my first intervention in here, hope someone will find it somewhat usefull :relieved: 
cheers and thanks again for this really great idea and feature :triumph: :star_struck:

**Comment by fdncred** [comment]
Thanks for the feedback @amtoine. We're trying hard on the feature because it's pretty mind-blowing and it inspires creative thinking about nushell, IMO.

> more vim-bindings and especially motions

We're trying to move away from vim-like behaviors right now. There was some discussion on the core team of not liking the vim-like modal type behavior. This is the reasons for some of my suggestions above.

> and finally a page about the explore tool would be great, for instance in the book

agreed. I tried to put some information in the [PR body](https://github.com/nushell/nushell/pull/6984) with a bunch of screenshots but we definitely need documentation. PRs are welcome but those docs will need to change too as we change the behavior.

**Comment by yingzhu146** [comment]
Two inputs from my side:

# 1) make vim mode a first class citizen - there is clear precedence and evidence of interest via visidata (6.2k GH stars)
> > more vim-bindings and especially motions
> 
> We're trying to move away from vim-like behaviors right now. There was some discussion on the core team of not liking the vim-like modal type behavior. This is the reasons for some of my suggestions above.

Second @amtoine - would prefer more vim modes. I think a great example of where this can develop into is [visidata](https://github.com/saulpw/visidata)

Completely understand where you're coming from - increase discoverability + reduce learning curve. But "moving away from" is the wrong way to go about this IMO - more "adding a non-modal emacs mode". I think the popularity of visidata and the overlap of user personas with nushell give strong justification to double down on vim mode, and add emacs mode, too!

If users have "vim" mode in their nu config, by default current behavior should stay + more vim-like motions should be added.

I use visidata together with nushell on a daily basis - it's incredibly cool! So I have seen the light on where this is going.

```
pip3 install visidata
ls | to csv | vd --f csv -
```
is my current "version" of `explore`

# 2) eager mode / instant execute for `try`

Two examples
https://github.com/gpanders/ijq 
https://github.com/akavel/up with `--unsafe-full-throttle`

give this a try to test it (be super careful - it will execute on each key stroke)
`ls | to csv |   up --unsafe-full-throttle -c 'nu --stdin -c "from csv"' `
its super cool! 

Why is it super cool?
This UX pattern (i.e. `instant execution` that is insufficiently used often - learning curve depends on # of iterations required to get a new task & time per iteration - reducing time per iteration massively increases discoverability and reduces learning curve

Safety can be handled by adding more "ignore" commands (like `clear`) to the eager mode.

I'm thinking `:try-instant` or something could be added as command

**Comment by remmycat** [comment]
First of all, I encountered a panic:

one when resizing my terminal quickly when in `explore`:
```
thread 'main' panicked at 'overflow when subtracting durations', library/core/src/time.rs:930:9
```

But apart from that, I have used `explore` a bit over the last days and would like to give some feedback :]

### Color Scheme

When I opened the tool for the first time, the colors seemed all broken. The font was flickering between gray and white when scrolling and there were other artifacts. 

I believe the main problem here is that I'm using the macOS Terminal.App, which does not support RGB colors. I had no problems with this so far using nushell, the default themes only use named ANSI colours. (with the exception of `shape_garbage` for some reason?)

After adjusting the entire color scheme in the configuration, there was no more flickering or glitches.
I would very much expect the defaults to also be named ansi colors for `explore`.

I would also expect that the `explore` command takes any colors it needs from `$env.config.color_config` and that the `color_config` gets more fields for this where needed. Maybe in its own record inside the `color_config` if they are expected to stay very tool-specific.

### Keyboard operation

I like that people who get used to the vi keybindings can get lots of value out of using them in many different tools, but I think it is also benefitial to make nushell as intuitive as possible for people who _don't_ have much experience with any specific editor program.
(I also wish for nushell in general not to expect anyone to choose between `emacs` and `vi` edit styles with just expecting the user to know either one and what that means in detail.)

If this is a hard incompatibility, I personally think not using vi-like behaviours anywhere is the better choice.

Specificially for `explore` I think:
- it should _definitely_ allow quitting with Esc and/or ctrl-c
- it should guide the user how to get to the help screen somewhere, at least in the initial view, e.g. `Type :help for controls`
  - I think this was suggested somewhere else, so: Please do not just show help info the very first time someone opens the tool, people cannot even scroll up in the terminal to read that again when they need it. I'd much rather have a config like `show_banner` if it's too annoying after a while.

### Help screen

I find it a little confusing that this screen works just like the "normal" view and that you have to _use the tool_ to see all the info on _how to use the tool_…
But i guess one could argue that's true for nu help in general and the initial help screen shows enough to navigate to more info 🤔 

I think part of my confusion is due to it being "loaded as data", as if I entered a cell in the interactive mode.

I think I would personally prefer the help screen to be less "deep" and just show all the info for operating `explore` as text - clearly visually differentiated from the default view.

### Interactive mode

I find it a bit confusing that the interactive mode turns off when selecting a cell.
I would have expected it to stay on so I can quickly navigate through the data interactively, instead of having to press `i` again and again.

### Navigating data in general

I would really like to somehow see a form of state of what I'm currently viewing.

We have a little bit of that in the bar at the bottom where it currently shows `information`
<img width="777" alt="Screenshot 2022-12-04 at 00 41 19" src="https://user-images.githubusercontent.com/3317423/205467163-080d31fc-3485-4c72-a1c8-eaeef842a9e4.png">

I think one way to achieve a better orientation would be some kind of "full path" in that place?

e.g. when running `sys | explore`, the bar could initially say
```
$in
```

Then I navigate to the fields `host` and sessions (via `i`) and the bar shows:
```
$in.host.sessions
```

I select the cell at column `name` in the fourth row:
```
$in.host.sessions.3.name
```

I go back via Esc:
```
$in.host.sessions
```

You could also show `:nu` and other commands, e.g. when writing `:nu where type == 'file'` the bar could say
```
$in.configuration | where type === 'file'
```

That way you would always know where you are. It might also help with the worries on Esc closing `explore`, because you would be more sure of when you are back at the initial screen.

If it gets too long the path would need to be truncated (on the left?)

The ideas I mentioned in the PR originally (that partially lead to the `:try` mode) where based on this, and being able to edit this line and add your own commands to the "path".

I'm wondering if there is still a possiblity to combine the normal view with `:try` mode in this way. While the bar at the bottom wouldn't be great visually for editing the line, i could imagine to fluently switch between the normal and `:try` mode, while showing the same data. (The input field in `:try` would then be prefilled with the "path")

### Try mode

I think this was partially based on something I suggested and it's really nice that you implemented this, @zhiburt, thank you!

Apart from what I already wrote about maybe fluently switching or combining the modes more, one thing that confused me a lot when opening `:try`, is that it shows an empty view.
I would have expected that it should at least show `$in` if there was input given to `explore`?
(I'm really glad `$in` works there now! After the initial PR got merged it gave me a panic)


---

Aaanyway, that has been a lot of rambling from a rando 😅 - I would be really interested on your thoughts on this though! I could also imagine trying to help implement some parts with PRs, though trying to familiarise myself with the `explore` code has been rather overwhelming so far.

(Note: Unfortunately I'm not able to have very active conversations on GitHub on most days, so I might seem "unresponsive" if there is much back and forth)

**Comment by fdncred** [comment]
There were mostly addressed with #7339 so let's close this. we have more ideas too.
